File: Backend/controllers/authController.js
/**
 * Authentication Controller
 * Handles all authentication-related operations
 */

const jwt = require('jsonwebtoken');
const crypto = require('crypto');
const User = require('../models/User');
const Authority = require('../models/Authority');

// API Configuration
const API_HOST = process.env.API_HOST || '192.168.100.35';
const API_PORT = process.env.API_PORT || '5000';
const API_BASE = `http://${API_HOST}:${API_PORT}`;

// Generate JWT Token
const generateToken = (id, role) => {
  return jwt.sign(
    { id, role },
    process.env.JWT_SECRET,
    { expiresIn: process.env.JWT_EXPIRE || '7d' }
  );
};

// Store reset tokens (use Redis in production)
const resetTokens = new Map();

/**
 * Helper: Generate RTSP URL from camera details
 * Keeps DB schema the same while making signup user-friendly.
 */
const generateRtspUrlFromCamera = ({
  ip,
  username,
  password,
  port = 554,
  brand,
  path
}) => {
  if (!ip) return null;

  const normalizedBrand = (brand || '').toLowerCase();
  let streamPath = path;

  if (!streamPath) {
    if (normalizedBrand.includes('hikvision')) {
      streamPath = '/Streaming/Channels/101';
    } else if (normalizedBrand.includes('dahua')) {
      streamPath = '/cam/realmonitor?channel=1&subtype=1';
    } else if (normalizedBrand.includes('meari')) {
      streamPath = '/live';
    } else {
      streamPath = '/cam/realmonitor?channel=1&subtype=1';
    }
  }

  // Meari cameras typically use port 8554 for RTSP
  let finalPort = port || 554;
  if (normalizedBrand.includes('meari') && !port) {
    finalPort = 8554;
  }

  // Build RTSP URL — credentials are optional
  if (username && password) {
    return `rtsp://${encodeURIComponent(username)}:${encodeURIComponent(password)}@${ip}:${finalPort}${streamPath}`;
  }
  return `rtsp://${ip}:${finalPort}${streamPath}`;
};

/**
 * @desc    Register a new user
 * @route   POST /api/auth/signup/user
 * @access  Public
 */
const registerUser = async (req, res, next) => {
  try {
    const {
      name,
      email,
      phone,
      password,
      cctvName,
      rtspUrl,
      location,
      // New camera-friendly fields (used only to generate RTSP)
      cameraIp,
      cameraUsername,
      cameraPassword,
      cameraPort,
      cameraBrand,
      cameraPath,
    } = req.body;

    // Check if user exists
    const existingUser = await User.findOne({ email: email.toLowerCase() });
    if (existingUser) {
      return res.status(400).json({
        success: false,
        error: 'User with this email already exists'
      });
    }

    // Determine final RTSP URL.
    // Priority: explicit rtspUrl (for backward compatibility),
    // otherwise generate from camera fields.
    let finalRtspUrl = rtspUrl;

    if (!finalRtspUrl) {
      finalRtspUrl = generateRtspUrlFromCamera({
        ip: cameraIp,
        username: cameraUsername,
        password: cameraPassword,
        port: cameraPort,
        brand: cameraBrand,
        path: cameraPath,
      });
    }

    if (!finalRtspUrl) {
      return res.status(400).json({
        success: false,
        error: 'Unable to generate RTSP URL from provided camera details',
      });
    }

    // Create user (password hashing handled by model)
    const user = await User.create({
      name,
      email: email.toLowerCase(),
      phone,
      password,
      cctvName,
      rtspUrl: finalRtspUrl,
      location
    });

    res.status(201).json({
      success: true,
      message: 'User registered successfully',
      data: {
        id: user._id,
        name: user.name,
        email: user.email
      }
    });
  } catch (error) {
    next(error);
  }
};

/**
 * @desc    Register a new authority
 * @route   POST /api/auth/signup/authority
 * @access  Public
 */
const registerAuthority = async (req, res, next) => {
  try {
    const { name, email, officerId, stationName, password } = req.body;

    // Check if authority exists
    const existingAuthority = await Authority.findOne({ 
      $or: [
        { email: email.toLowerCase() },
        { officerId: officerId.toUpperCase() }
      ]
    });
    
    if (existingAuthority) {
      return res.status(400).json({
        success: false,
        error: 'Authority with this email or officer ID already exists'
      });
    }

    // Create authority
    const authority = await Authority.create({
      name,
      email: email.toLowerCase(),
      officerId: officerId.toUpperCase(),
      stationName,
      password
    });

    res.status(201).json({
      success: true,
      message: 'Authority registered successfully',
      data: {
        id: authority._id,
        name: authority.name,
        email: authority.email,
        officerId: authority.officerId
      }
    });
  } catch (error) {
    next(error);
  }
};

/**
 * @desc    Login user or authority
 * @route   POST /api/auth/login
 * @access  Public
 */
const login = async (req, res, next) => {
  try {
    const { email, password } = req.body;
    const emailLower = email.toLowerCase();

    // Try to find user first
    let user = await User.findOne({ email: emailLower }).select('+password');
    let role = 'user';
    let camera = null;

    if (!user) {
      // Try to find authority
      user = await Authority.findOne({ email: emailLower }).select('+password');
      role = user?.role || 'authority';
    } else {
      // Build camera info for user
      camera = {
        camera_name: user.cctvName,
        stream_url: `${API_BASE}/streams/stream.m3u8`,
        location: user.location,
        rtsp_url: user.rtspUrl
      };
    }

    if (!user) {
      return res.status(401).json({
        success: false,
        error: 'Invalid email or password'
      });
    }

    if (!user.isActive) {
      return res.status(401).json({
        success: false,
        error: 'Account is deactivated. Please contact support.'
      });
    }

    // Check password
    const isMatch = await user.comparePassword(password);
    if (!isMatch) {
      return res.status(401).json({
        success: false,
        error: 'Invalid email or password'
      });
    }

    // Update last login
    user.lastLogin = new Date();
    await user.save({ validateBeforeSave: false });

    // Generate token
    const token = generateToken(user._id, role);

    // Build response
    const userData = {
      id: user._id,
      _id: user._id,
      name: user.name,
      email: user.email,
      role,
      ...(camera && { camera }),
      ...(role === 'user' && {
        phone: user.phone,
        cctvName: user.cctvName,
        location: user.location
      }),
      ...(role !== 'user' && {
        officerId: user.officerId,
        stationName: user.stationName,
        department: user.department,
        isVerified: user.isVerified
      })
    };

    res.json({
      success: true,
      token,
      user: userData,
      role
    });
  } catch (error) {
    next(error);
  }
};

/**
 * @desc    Get current user profile
 * @route   GET /api/auth/me
 * @access  Private
 */
const getProfile = async (req, res, next) => {
  try {
    res.json({
      success: true,
      data: req.user
    });
  } catch (error) {
    next(error);
  }
};

/**
 * @desc    Update user profile
 * @route   PUT /api/auth/profile
 * @access  Private
 */
const updateProfile = async (req, res, next) => {
  try {
    const allowedUpdates = ['name', 'phone', 'cctvName', 'rtspUrl', 'location'];
    const updates = {};

    allowedUpdates.forEach(field => {
      if (req.body[field] !== undefined) {
        updates[field] = req.body[field];
      }
    });

    const user = await User.findByIdAndUpdate(
      req.user._id,
      updates,
      { new: true, runValidators: true }
    );

    res.json({
      success: true,
      message: 'Profile updated successfully',
      data: user
    });
  } catch (error) {
    next(error);
  }
};

/**
 * @desc    Change password
 * @route   PUT /api/auth/change-password
 * @access  Private
 */
const changePassword = async (req, res, next) => {
  try {
    const { currentPassword, newPassword } = req.body;

    // Get user with password
    const user = await User.findById(req.user._id).select('+password') ||
                 await Authority.findById(req.user._id).select('+password');

    if (!user) {
      return res.status(404).json({
        success: false,
        error: 'User not found'
      });
    }

    // Check current password
    const isMatch = await user.comparePassword(currentPassword);
    if (!isMatch) {
      return res.status(400).json({
        success: false,
        error: 'Current password is incorrect'
      });
    }

    // Update password
    user.password = newPassword;
    await user.save();

    res.json({
      success: true,
      message: 'Password changed successfully'
    });
  } catch (error) {
    next(error);
  }
};

/**
 * @desc    Forgot password - request reset
 * @route   POST /api/auth/forgot-password
 * @access  Public
 */
const forgotPassword = async (req, res, next) => {
  try {
    const { email } = req.body;
    const emailLower = email.toLowerCase();

    // Find user or authority
    let user = await User.findOne({ email: emailLower });
    let userType = 'user';
    
    if (!user) {
      user = await Authority.findOne({ email: emailLower });
      userType = 'authority';
    }

    // Always return success to prevent email enumeration
    if (!user) {
      return res.json({
        success: true,
        message: 'If an account exists with this email, a reset link has been sent.'
      });
    }

    // Generate reset token
    const resetToken = crypto.randomBytes(32).toString('hex');
    const resetTokenExpiry = Date.now() + 30 * 60 * 1000; // 30 minutes

    // Store token
    resetTokens.set(resetToken, {
      email: emailLower,
      userType,
      expiry: resetTokenExpiry
    });

    // Log reset link (in production, send email)
    console.log(`Password reset token for ${email}: ${resetToken}`);
    console.log(`Reset link: ${API_BASE}/api/auth/reset-password/${resetToken}`);

    res.json({
      success: true,
      message: 'Password reset link has been sent to your email',
      ...(process.env.NODE_ENV !== 'production' && { resetToken })
    });
  } catch (error) {
    next(error);
  }
};

/**
 * @desc    Reset password with token
 * @route   POST /api/auth/reset-password/:token
 * @access  Public
 */
const resetPassword = async (req, res, next) => {
  try {
    const { token } = req.params;
    const { password } = req.body;

    // Validate token
    const tokenData = resetTokens.get(token);
    if (!tokenData) {
      return res.status(400).json({
        success: false,
        error: 'Invalid or expired reset token'
      });
    }

    // Check expiry
    if (Date.now() > tokenData.expiry) {
      resetTokens.delete(token);
      return res.status(400).json({
        success: false,
        error: 'Reset token has expired'
      });
    }

    // Find and update user
    let user;
    if (tokenData.userType === 'authority') {
      user = await Authority.findOne({ email: tokenData.email });
    } else {
      user = await User.findOne({ email: tokenData.email });
    }

    if (!user) {
      return res.status(400).json({
        success: false,
        error: 'User not found'
      });
    }

    // Update password
    user.password = password;
    await user.save();

    // Delete used token
    resetTokens.delete(token);

    res.json({
      success: true,
      message: 'Password reset successfully'
    });
  } catch (error) {
    next(error);
  }
};

/**
 * @desc    Logout user (client-side token deletion)
 * @route   POST /api/auth/logout
 * @access  Private
 */
const logout = async (req, res, next) => {
  try {
    // Stop AIService detection for this user session
    const axios = require('axios');
    const AI_SERVICE_URL = process.env.AI_SERVICE_URL || 'http://localhost:8000';
    try {
      await axios.post(`${AI_SERVICE_URL}/stop-detection`, {}, { timeout: 5000 });
    } catch (err) {
      // Log but don't block logout if AIService is unreachable
      console.error('Failed to notify AIService to stop detection:', err.message);
    }
    res.json({
      success: true,
      message: 'Logged out successfully and AIService stopped.'
    });
  } catch (error) {
    next(error);
  }
};

/**
 * @desc    Google OAuth callback handler
 * @route   GET /api/auth/google/callback
 * @access  Public
 */
const googleCallback = async (req, res) => {
  try {
    const user = req.user;
    const token = generateToken(user._id, user.role);

    // Update last login
    user.lastLogin = new Date();
    await user.save({ validateBeforeSave: false });

    // Build user data for the frontend
    const userData = {
      _id: user._id,
      name: user.name,
      email: user.email,
      role: user.role,
      avatar: user.avatar,
      cctvName: user.cctvName,
      rtspUrl: user.rtspUrl,
      location: user.location,
    };

    const FRONTEND_URL = process.env.FRONTEND_URL || 'http://localhost:3000';
    const params = new URLSearchParams({
      token,
      user: JSON.stringify(userData),
    });
    res.redirect(`${FRONTEND_URL}/login?${params.toString()}`);
  } catch (error) {
    console.error('Google callback error:', error);
    const FRONTEND_URL = process.env.FRONTEND_URL || 'http://localhost:3000';
    res.redirect(`${FRONTEND_URL}/login?error=google_auth_failed`);
  }
};

module.exports = {
  registerUser,
  registerAuthority,
  login,
  getProfile,
  updateProfile,
  changePassword,
  forgotPassword,
  resetPassword,
  logout,
  googleCallback,
  // Exported for reuse in other controllers (e.g., camerasController)
  generateRtspUrlFromCamera,
};

File: Backend.NET/src/AttackShield.API/Controllers/AuthController.cs
using System.Collections.Concurrent;
using AttackShield.Core.DTOs;
using AttackShield.Core.Entities;
using AttackShield.Core.Interfaces;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace AttackShield.Api.Controllers;

/// <summary>
/// Authentication and profile endpoints. Ported from the Node authController.
/// Users and authorities share the login/reset flow; role is inferred by which
/// collection the email is found in. Split across partial files: this one holds
/// signup + login; the rest (profile, password, OAuth) live in the .Account file.
/// </summary>
[Route("api/auth")]
public sealed partial class AuthController : ApiControllerBase
{
    private readonly IUserRepository _users;
    private readonly IAuthorityRepository _authorities;
    private readonly IPasswordHasher _hasher;
    private readonly IJwtTokenService _jwt;
    private readonly IRtspUrlBuilder _rtsp;
    private readonly IAiServiceClient _ai;
    private readonly IConfiguration _config;

    // In-memory password-reset tokens (the original used a Map; Redis in prod).
    private static readonly ConcurrentDictionary<string, ResetTokenData> ResetTokens = new();

    public AuthController(
        IUserRepository users,
        IAuthorityRepository authorities,
        IPasswordHasher hasher,
        IJwtTokenService jwt,
        IRtspUrlBuilder rtsp,
        IAiServiceClient ai,
        IConfiguration config)
    {
        _users = users;
        _authorities = authorities;
        _hasher = hasher;
        _jwt = jwt;
        _rtsp = rtsp;
        _ai = ai;
        _config = config;
    }

    private string ApiBase => _config["Backend:BaseUrl"] ?? "http://localhost:5000";
    private string FrontendUrl => _config["Frontend:BaseUrl"] ?? "http://localhost:3000";

    [HttpPost("signup/user")]
    [AllowAnonymous]
    public async Task<IActionResult> RegisterUser([FromBody] RegisterUserRequest req, CancellationToken ct)
    {
        if (string.IsNullOrWhiteSpace(req.Email) || string.IsNullOrWhiteSpace(req.Password) || string.IsNullOrWhiteSpace(req.Name))
            return Fail("Name, email and password are required");

        var existing = await _users.GetByEmailAsync(req.Email, ct);
        if (existing is not null)
            return Fail("User with this email already exists");

        var finalRtsp = req.RtspUrl;
        if (string.IsNullOrWhiteSpace(finalRtsp))
            finalRtsp = _rtsp.Build(req.CameraIp, req.CameraUsername, req.CameraPassword, req.CameraPort, req.CameraBrand, req.CameraPath);

        if (string.IsNullOrWhiteSpace(finalRtsp))
            return Fail("Unable to generate RTSP URL from provided camera details");

        var user = new User
        {
            Name = req.Name,
            Email = req.Email.ToLowerInvariant(),
            Phone = req.Phone,
            Password = _hasher.Hash(req.Password),
            CctvName = req.CctvName,
            RtspUrl = finalRtsp,
            Location = req.Location,
        };
        await _users.InsertAsync(user, ct);

        return StatusCode(201, new
        {
            success = true,
            message = "User registered successfully",
            data = new { id = user.Id, name = user.Name, email = user.Email },
        });
    }

    [HttpPost("signup/authority")]
    [AllowAnonymous]
    public async Task<IActionResult> RegisterAuthority([FromBody] RegisterAuthorityRequest req, CancellationToken ct)
    {
        if (string.IsNullOrWhiteSpace(req.Email) || string.IsNullOrWhiteSpace(req.Password)
            || string.IsNullOrWhiteSpace(req.OfficerId) || string.IsNullOrWhiteSpace(req.Name)
            || string.IsNullOrWhiteSpace(req.StationName))
            return Fail("All fields are required");

        var existing = await _authorities.GetByEmailOrOfficerIdAsync(req.Email, req.OfficerId, ct);
        if (existing is not null)
            return Fail("Authority with this email or officer ID already exists");

        var authority = new Authority
        {
            Name = req.Name,
            Email = req.Email.ToLowerInvariant(),
            OfficerId = req.OfficerId.ToUpperInvariant(),
            StationName = req.StationName,
            Password = _hasher.Hash(req.Password),
        };
        await _authorities.InsertAsync(authority, ct);

        return StatusCode(201, new
        {
            success = true,
            message = "Authority registered successfully",
            data = new { id = authority.Id, name = authority.Name, email = authority.Email, officerId = authority.OfficerId },
        });
    }

    [HttpPost("login")]
    [AllowAnonymous]
    public async Task<IActionResult> Login([FromBody] LoginRequest req, CancellationToken ct)
    {
        if (string.IsNullOrWhiteSpace(req.Email) || string.IsNullOrWhiteSpace(req.Password))
            return Fail("Email and password are required");

        var emailLower = req.Email.ToLowerInvariant();
        var user = await _users.GetByEmailAsync(emailLower, ct);

        if (user is null)
        {
            // Fall back to the authorities collection.
            var authority = await _authorities.GetByEmailAsync(emailLower, ct);
            if (authority is null)
                return Fail("Invalid email or password", 401);
            if (!authority.IsActive)
                return Fail("Account is deactivated. Please contact support.", 401);
            if (string.IsNullOrEmpty(authority.Password) || !_hasher.Verify(req.Password, authority.Password))
                return Fail("Invalid email or password", 401);

            authority.LastLogin = DateTime.UtcNow;
            await _authorities.UpdateAsync(authority.Id!, authority, ct);
            var authToken = _jwt.GenerateToken(authority.Id!, authority.Role);

            return Ok(new
            {
                success = true,
                token = authToken,
                user = new
                {
                    id = authority.Id,
                    _id = authority.Id,
                    name = authority.Name,
                    email = authority.Email,
                    role = authority.Role,
                    officerId = authority.OfficerId,
                    stationName = authority.StationName,
                    department = authority.Department,
                    isVerified = authority.IsVerified,
                },
                role = authority.Role,
            });
        }

        if (!user.IsActive)
            return Fail("Account is deactivated. Please contact support.", 401);
        if (string.IsNullOrEmpty(user.Password) || !_hasher.Verify(req.Password, user.Password))
            return Fail("Invalid email or password", 401);

        user.LastLogin = DateTime.UtcNow;
        await _users.UpdateAsync(user.Id!, user, ct);
        var token = _jwt.GenerateToken(user.Id!, "user");

        var camera = new
        {
            camera_name = user.CctvName,
            stream_url = $"{ApiBase}/streams/stream.m3u8",
            location = user.Location,
            rtsp_url = user.RtspUrl,
        };

        return Ok(new
        {
            success = true,
            token,
            user = new
            {
                id = user.Id,
                _id = user.Id,
                name = user.Name,
                email = user.Email,
                role = "user",
                camera,
                phone = user.Phone,
                cctvName = user.CctvName,
                location = user.Location,
            },
            role = "user",
        });
    }

    /// <summary>Reset-token record kept in the in-memory store.</summary>
    private sealed record ResetTokenData(string Email, string UserType, DateTime Expiry);
}
